test/e2e/self_hosted.go

```diff
@@ -265,3 +386,17 @@ func SelfHostedSpec(ctx context.Context, inputGetter func() SelfHostedSpecInput)
 		dumpSpecResourcesAndCleanup(ctx, specName, input.BootstrapClusterProxy, input.ArtifactFolder, namespace, cancelWatches, clusterResources.Cluster, input.E2EConfig.GetIntervals, input.SkipCleanup)
 	})
 }
+
+func hasProvider(ctx context.Context, c client.Client, providerName string) bool {
+	providerList := clusterctlv1.ProviderList{}
+	Eventually(func() error {
+		return c.List(ctx, &providerList)
+	}, "1m", "5s").Should(Succeed(), "Failed to list the Providers")
+
+	for _, provider := range providerList.Items {
+		if provider.GetName() == providerName {
+			return true
+		}
+	}
+	return false
+}
```

Comment:
The problem is that we have to read the cluster template / ClusterClass YAML for that before deploying it. Afterwards it's too late.

I thought the trade-off could be okay, given that the Docker env vars are scoped to DOCKER

Reply:
Ok let's just make this more explicit where the func is called, something like

```
// In case the infrastructure-docker provider is installed, ensure to add the preload images variable to load the
// controller images into the nodes.
// NOTE: we are checking the bootstrap cluster and assuming the workload cluster will be on the same
// infrastructure provider. Also, given that we use it to set a variable, then it is up to cluster templates to use it or not.
hasDockerInfrastructureProvider := hasProvider(ctx, input.BootstrapClusterProxy.GetClient(), "infrastructure-docker")
```

Reply:
added